### PR TITLE
Added 'resize' operation to BoundedQueue

### PR DESCRIPTION
**Pull request by jpkrohling**
Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

## Which problem is this PR solving?
- This PR is a prerequisite for #943

## Short description of the changes
- This adds a new `Resize(int)` method to the `BoundedQueue`

### CODE REVIEW

**Comment by yurishkuro** [comment]
how is it supposed to be used? who would call the Resize()? Isn't it possible to just pre-determine the queue size upfront?

**Comment by jpkrohling** [comment]
> how is it supposed to be used? who would call the Resize()? 

A change to the span processor will make it keep track of the number of spans and the total bytes it has seen during the process lifetime. Once enough spans have been recorded, the span processor starts to calculate what's the ideal queue size based on the average number of spans and available memory. It can then decide to trigger a `Resize()` operation.

I have a local branch that does the above, but I thought it would be cleaner to separate those two features into separate PRs.

> Isn't it possible to just pre-determine the queue size upfront?

When we know what's the expected average span size, yes. But I've not heard of any cases where people know in advance what's the average span size, making the queue size hard to calculate.

**Comment by jpkrohling** [comment]
I just added a benchmark to the tests, to make ensure that I'm not bringing performance problems to the queue. Here are the results comparing master with this PR, which shows no performance degradation for current users of the queue (unless I'm making something really wrong with the benchmark):

```
$ git branch --show-current 
master

$ go test -benchmem -run=^$ ./pkg/queue -bench "^(BenchmarkBoundedQueue)$"
goos: linux
goarch: amd64
pkg: github.com/jaegertracing/jaeger/pkg/queue
BenchmarkBoundedQueue-8   	19407266	        52.2 ns/op	       8 B/op	       1 allocs/op
PASS
ok  	github.com/jaegertracing/jaeger/pkg/queue	1.081s

$ git branch --show-current 
Resize-bounded-queue

$ go test -benchmem -run=^$ ./pkg/queue -bench "^(BenchmarkBoundedQueue)$"
goos: linux
goarch: amd64
pkg: github.com/jaegertracing/jaeger/pkg/queue
BenchmarkBoundedQueue-8   	20110519	        51.4 ns/op	       8 B/op	       1 allocs/op
PASS
ok  	github.com/jaegertracing/jaeger/pkg/queue	1.099s
```

**Comment by yurishkuro** [comment]
I am skeptical about relying on the average span size, because the actual size differences can be very large, 2-3 orders of magnitude, e.g a basic span vs. span with a stack trace vs. span with request payload added to the logs. I think if we want to go by memory size, we should go by the actual size of spans in the queue. The proto/domain model already implements Size() method iirc, which could be used as an approximation of the actual memory size, and we would need to re-implement the queue using an explicit buffer rather than a bound channel (might be worth looking around for an existing implementation).

**Comment by jpkrohling** [comment]
> the actual size differences can be very large, 2-3 orders of magnitude, e.g a basic span vs. span with a stack trace vs. span with request payload added to the logs

You are right, but in reality, this wouldn't be a big problem. The queue sizes will typically be big enough to make outliers non-relevant. In [my working branch](https://github.com/jpkrohling/jaeger/tree/943-DynQueueSize) for #943, spans generated by `tracegen` have an average of 1.5KiB, which means a queue size of about 700k items at 1GiB. And even if spans start being bigger all of a sudden, the [periodic background process](https://github.com/jpkrohling/jaeger/blob/bac686546a9cc8bae07c016e9ad05bdfd8dd52d2/cmd/collector/app/span_processor.go#L205-L209) will readjust the queue size as long as the difference is bigger [than a few % points](https://github.com/jpkrohling/jaeger/blob/bac686546a9cc8bae07c016e9ad05bdfd8dd52d2/cmd/collector/app/span_processor.go#L258-L263). Note that the "memory" aspect is an approximation: there shouldn't be a crash if we end up using more memory for a short period of time. And most of the times, the queue should be empty anyway ;-)

I would give this simple approach a try first and see how it works in the real world.

**Comment by jpkrohling** [comment]
@yurishkuro any feedback on this one?

File: pkg/queue/bounded_queue.go [hunk 1]
```diff
@@ -97,17 +105,52 @@ func (q *BoundedQueue) Stop() {
...
+	// if the capacity size hasn't changed, there's no need to resize
+	if cap(q.items) == newCap {
+		return false
+	}
+
+	// we need exclusive rw access to the internal queue
+	q.resizeLock.Lock()
+	defer q.resizeLock.Unlock()
+
+	new := make(chan interface{}, newCap)
```
Comment (yurishkuro) [inline pkg/queue/bounded_queue.go]: * what if newCap is less than currentCap and current size? This will probably deadlock
* please don't use `new` keyword as var name

Comment (jpkrohling) [reply pkg/queue/bounded_queue.go]: Funny, I thought I had added a test for the downscale, but looks like I haven't. Will add one.

**Comment by yurishkuro** [comment]
I wonder if it's possible to avoid the mutex by approaching this differently. Go routines are cheap to start, so rather than copying the messages into another queue, we could start another set of workers reading from a new channel, close the original channel and let those old workers drain it naturally and exit.

**Comment by jpkrohling** [comment]
That could work, let me give it a try.

**Comment by jpkrohling** [comment]
PR updated, to use a new channel for the resized queue.

**Comment by jpkrohling** [comment]
@yurishkuro would you please re-review?

File: pkg/queue/bounded_queue_test.go [hunk 2]
```diff
@@ -156,3 +157,173 @@ func (s *consumerState) assertConsumed(expected map[string]bool) {
...
+	})
+
+	var firstConsumer, secondConsumer, releaseConsumers sync.WaitGroup
+	firstConsumer.Add(1)
+	secondConsumer.Add(1)
+	releaseConsumers.Add(1)
+
+	released, resized := false, false
+	q.StartConsumers(1, func(item interface{}) {
+		if !resized { // we'll have a second consumer once the queue is resized
```
Comment (pavolloffay) [inline pkg/queue/bounded_queue_test.go]: I didn't find the second consumer

File: pkg/queue/bounded_queue_test.go [hunk 3]
```diff
@@ -156,3 +157,173 @@ func (s *consumerState) assertConsumed(expected map[string]bool) {
...
+	})
+
+	var consumer, releaseConsumers sync.WaitGroup
+	consumer.Add(1)
+	releaseConsumers.Add(1)
+
+	released := false
+	q.StartConsumers(1, func(item interface{}) {
+		// once we release the lock, we might end up with multiple calls to reach this
+		if !released {
```
Comment (pavolloffay) [inline pkg/queue/bounded_queue_test.go]: Released does not seem to be needed

File: pkg/queue/bounded_queue_test.go [hunk 4]
```diff
@@ -156,3 +157,173 @@ func (s *consumerState) assertConsumed(expected map[string]bool) {
...
+	assert.EqualValues(t, 4, q.Capacity())
+	assert.EqualValues(t, q.Capacity(), q.Size())
+	assert.EqualValues(t, q.Capacity(), len(*q.items))
+
+	assert.True(t, q.Resize(2))
+	assert.False(t, q.Produce("f")) // dropped
+
+	assert.EqualValues(t, 2, q.Capacity())
+	assert.EqualValues(t, 4, q.Size())      // the queue will eventually drain, but it will live for a while over capacity
+	assert.EqualValues(t, 0, len(*q.items)) // the new queue is empty, as the old queue is still full and over capacity
```
Comment (pavolloffay) [inline pkg/queue/bounded_queue_test.go]: Shouldn't we keep accepting here? There are free spots in the queue.

File: pkg/queue/bounded_queue.go [hunk 5]
```diff
@@ -79,11 +89,23 @@ func (q *BoundedQueue) Produce(item interface{}) bool {
 		q.onDroppedItem(item)
 		return false
 	}
+
+	// we might have two concurrent backing queues at the moment
+	// their combined size is stored in q.size, and their combined capacity
+	// should match the capacity of the new queue
+	if atomic.LoadInt32(&q.size) >= int32(q.Capacity()) && q.Capacity() > 0 {
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply pkg/queue/bounded_queue.go]: There are consumers of this queue that are not setting the queue size. IIRC, one of them is in the span processor tests. Agree with your earlier suggestion though.

File: pkg/queue/bounded_queue_test.go (hunk 2 above)
Comment (jpkrohling) [reply pkg/queue/bounded_queue_test.go]: Around line 200:

https://github.com/jaegertracing/jaeger/blob/a7819b895930fc85727750f2a00cf5b21bc8cbb7/pkg/queue/bounded_queue_test.go#L197-L200

File: pkg/queue/bounded_queue_test.go (hunk 3 above)
Comment (jpkrohling) [reply pkg/queue/bounded_queue_test.go]: Why's that? Note that there's quite a bunch of go funcs just waiting for this lock to be released to start calling consumers with other items. If we have *not* released, we mark one step down in the work group. If we have released the consumer, we do not want to count down further, as we'd reach a negative count.

File: pkg/queue/bounded_queue_test.go (hunk 4 above)
Comment (jpkrohling) [reply pkg/queue/bounded_queue_test.go]: There are free spots in the individual channel, but if the queue size was set by the user to 3000, it's wrong to have 6000 items in the two queues combined.

File: pkg/queue/bounded_queue.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (pavolloffay) [reply pkg/queue/bounded_queue.go]: I see there is a test which sets the size to `0`

File: pkg/queue/bounded_queue_test.go (hunk 3 above)
Comment (pavolloffay) [reply pkg/queue/bounded_queue_test.go]: I mean you can remove this parameter altogether. It's not changed during the test - only at the end which does not affect tests and the default value satisfy this if statement.

Comment (pavolloffay) [reply pkg/queue/bounded_queue_test.go]: Removal of this variable does not have any effect on the tests. It seems redundant

Comment (jpkrohling) [reply pkg/queue/bounded_queue_test.go]: It needs to be there, otherwise the test will intermittently fail due to "negative counts" in the work group.

**Comment by jpkrohling** [comment]
Thanks!

File: pkg/queue/bounded_queue.go [hunk 6]
```diff
@@ -29,46 +30,55 @@ import (
...
+						// channel closed, finish worker
+						return
+					}
 				case <-q.stopCh:
+					// the whole queue is closing, finish worker
 					return
 				}
 			}
-		}()
+		}(*q.items)
```
Comment (yurishkuro) [inline pkg/queue/bounded_queue.go]: Accessing the pointer like this creates an unnecessary race condition where some workers can theoretically start against a different queue. A simpler solution is to dereference the field outside of the loop.

File: pkg/queue/bounded_queue.go [hunk 7]
```diff
@@ -29,46 +30,55 @@ import (
...
 // the items from the top of the queue until its size drops back to maxSize
 type BoundedQueue struct {
-	capacity      int
 	size          int32
 	onDroppedItem func(item interface{})
-	items         chan interface{}
+	items         *chan interface{}
 	stopCh        chan struct{}
 	stopWG        sync.WaitGroup
 	stopped       int32
```
Comment (yurishkuro) [inline pkg/queue/bounded_queue.go]: we should move this field to the top to ensure word alignment

File: pkg/queue/bounded_queue.go [hunk 8]
```diff
@@ -29,46 +30,55 @@ import (
 // channels, with a special Reaper goroutine that wakes up when the queue is full and consumers
 // the items from the top of the queue until its size drops back to maxSize
 type BoundedQueue struct {
-	capacity      int
 	size          int32
 	onDroppedItem func(item interface{})
-	items         chan interface{}
+	items         *chan interface{}
```
Comment (yurishkuro) [inline pkg/queue/bounded_queue.go]: This is being updated with a CAS in one place but accessed directly in other places, which is technically not thread safe. I would rather we used a normal mutex or `atomic.Value`.

File: pkg/queue/bounded_queue.go [hunk 9]
```diff
@@ -79,11 +89,23 @@ func (q *BoundedQueue) Produce(item interface{}) bool {
...
+
+	// we might have two concurrent backing queues at the moment
+	// their combined size is stored in q.size, and their combined capacity
+	// should match the capacity of the new queue
+	if q.Size() >= q.Capacity() && q.Capacity() > 0 {
+		// current consumers of the queue (like tests) expect the queue capacity = 0 to work
+		// so, we don't drop items when the capacity is 0
+		q.onDroppedItem(item)
+		return false
+	}
```
Comment (yurishkuro) [inline pkg/queue/bounded_queue.go]: I don't understand this part. A second queue should never be the concern of Produce(), since it only operates on the _current_ queue. And why do we need capacity=0 for tests?

Comment (jpkrohling) [reply pkg/queue/bounded_queue.go]: > A second queue should never be the concern of Produce()

Correct, the comment is there only to make it explicit that there's no need to check two queues, as the state from both is handled at `q.Size()`. I can remove the comment if it's confusing.

> And why do we need capacity=0 for tests?

When no `Option.QueueSize()` is set and used in the span processor, the default queue size is 0 in tests like `cmd/collector/app/span_processor_test.go`. I changed the test to explicitly set a queue size, and removed the conditional here, as @pavolloffay had similar concerns.
